tests/test_plugin.py

```diff
@@ -61,20 +61,20 @@ class TestSplitToSuites:
             "least_duration",
             ["test_1", "test_2", "test_3", "test_4", "test_5", "test_6", "test_7", "test_8", "test_9", "test_10"],
         ),
-        (2, 1, "duration_based_chunks", ["test_1", "test_2", "test_3", "test_4", "test_5", "test_6", "test_7"]),
-        (2, 2, "duration_based_chunks", ["test_8", "test_9", "test_10"]),
+        (2, 1, "duration_based_chunks", ["test_1", "test_2", "test_3", "test_4", "test_5", "test_6"]),
+        (2, 2, "duration_based_chunks", ["test_7", "test_8", "test_9", "test_10"]),
         (2, 1, "least_duration", ["test_1", "test_3", "test_5", "test_7", "test_9"]),
         (2, 2, "least_duration", ["test_2", "test_4", "test_6", "test_8", "test_10"]),
         (3, 1, "duration_based_chunks", ["test_1", "test_2", "test_3", "test_4", "test_5"]),
-        (3, 2, "duration_based_chunks", ["test_6", "test_7", "test_8"]),
-        (3, 3, "duration_based_chunks", ["test_9", "test_10"]),
+        (3, 2, "duration_based_chunks", ["test_6", "test_7"]),
+        (3, 3, "duration_based_chunks", ["test_8", "test_9", "test_10"]),
         (3, 1, "least_duration", ["test_1", "test_4", "test_7", "test_10"]),
         (3, 2, "least_duration", ["test_2", "test_5", "test_8"]),
         (3, 3, "least_duration", ["test_3", "test_6", "test_9"]),
-        (4, 1, "duration_based_chunks", ["test_1", "test_2", "test_3", "test_4"]),
-        (4, 2, "duration_based_chunks", ["test_5", "test_6", "test_7"]),
-        (4, 3, "duration_based_chunks", ["test_8", "test_9"]),
-        (4, 4, "duration_based_chunks", ["test_10"]),
+        (4, 1, "duration_based_chunks", ["test_1", "test_2", "test_3"]),
+        (4, 2, "duration_based_chunks", ["test_4", "test_5"]),
+        (4, 3, "duration_based_chunks", ["test_6"]),
+        (4, 4, "duration_based_chunks", ["test_7", "test_8", "test_9", "test_10"]),
```

Comment:
Doesn't this mean that the expected durations for each group will be:
* 1: 3s (1+1+1)
* 2: 2s (1+1)
* 3: 2s (2)
* 4: 8s (2+2+2+2)

If so, it doesn't feel right 🤔

Reply:
1-5 are 1 second tests, while 5-10 are 2 second tests, so I believe the timings are 

- 3 seconds
- 3 seconds 
- 2 seconds 
- 8 seconds

And yes this is weird, but it's also what you should expect if you want to *stop adding tests before going over the threshold* of 15/4=3,75 seconds, **and** keeping absolute order 🤷

Reply:
It's a very simple greedy algorithm; it's pretty much unavoidable that you're going to get strange far-from-optimal results like this.  The least_durations algorithm is much better and IMO should be the default and this algorithm should simply be removed.  But this algorithm does have the benefit of maintaining the sequence of tests, which I suppose could potentially be beneficial in some cases if there are ordering dependencies between tests.

Reply:
Actually, here's something to consider potentially: exceed the threshold, but keep track of a running total overage and use that when determining if you're over the threshold.  There are probably a few other heuristics that you could apply here to try to improve things a little bit.  For instance, something like this: ignore overage if the slot is empty and unconditionally assign, if it isn't empty then check to see if the resulting overage is less than half of the time required for that test, if so pack it, otherwise bump it to the next group.

Reply:
>  The least_durations algorithm is much better and IMO should be the default and this algorithm should simply be removed

I have a use case which requires this plugin to support maintaining the order of the tests, which is why we need to keep `duration_based_chunks` or something similar. However, I can consider changing the default but that'd probably bite some other users of this plugin as well so need to consider that carefully 🙂

Reply:
Like I said, I would vote for removing it, but if we can't remove it for legacy reasons then we might as well spend a few cycles making it a little bit better.  I believe the current least_durations algorithm also preserves the ordering of the tests, but doesn't keep them contiguous - i.e. it could put 1, 3, 5, 6 in one split and 2, 4, 7, 8 in another split, but it won't ever make a group of 1, 5, 3, 6.  In contrast, duration_based_chunks keeps them contiguous and could generate 1, 2, 3 and 4, 5, 6, 7, 8, or something along those lines.  But both can break if, say, test 4 requires test 3 to run beforehand, because either algorithm could place those tests in different groups.

Reply:
> But both can break if, say, test 4 requires test 3 to run beforehand, because either algorithm could place those tests in different groups.

True true, however duration_based_chunks most likely breaks less often, especially if there is a number of such dependencies between the tests 🙂

Reply:
Right, makes sense.  So, we should spend a few cycles improving the quality of the splitting in duration_based_chunks.  I think keeping track of the total "overage" while computing the splits should do quite a bit better.  If you wanted to do better than that, I think you might need to revisit the overall approach and perhaps come up with a less greedy algorithm.

Reply:
What ever we do here, at least my personal use case won't see much difference as it's thousands of tests (which each last 0-10 seconds) split into 3-5 groups, which is why I don't really have strong personal motivation behind improving the "general case". Thus, I think we could even go with this one from @sondrelg for now 🙂

Reply:
Feel free to jump in and make additional changes here - I don't have any need for this algo and was really only trying to help you out wrt. what was communicated in #27, so think I'll leave this as is 🙂

Reply:
Sounds like a reasonable plan to me - I'm also going to prefer the least_durations algorithm.  If I have a few minutes, perhaps I'll take a look at some tweaks to try to even things out a bit, but that can be done as a separate PR.

Reply:
I'll close this for now then 👍

Reply:
Well, without these changes, is duration_based_chunks still producing empty groups?  Or did that get fixed in a separate PR?

Reply:
OK, well my point was that even if the splitting isn't as perfect as it could be, at least this fix doesn't result in empty groups, so it should be merged and then possibly improved, instead of being closed.

Reply:
Sure thing. I've reopened it - feel free to make any additional changes you wish (cc @jerry-git) and merge. I'm on holiday now myself so likely won't be able to apply any code changes for a while.